pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,243 @@ package heal
+		pushFunc := healRequestFunc(ctx)
+		go u.listenToConnectionChanges(pushFunc, errCh)
+		u.initErr = <-errCh
+	})
+	// if initialization failed, then we want for all subsequent calls to Request() on this object to also fail
+	if u.initErr != nil {
+		return nil, u.initErr
+	}
+
+	// TODO what if this is a second call to Request() with this conn id?
+	//  will we receive a new event with this channel?
+	//  If not, then we will block on waiting and return an error eventually
+	verifySuccessCh, err := u.addToExpectedConnections(request.GetConnection())
+	if err != nil {
+		return nil, err
+	}
+
 	conn, err := next.Client(ctx).Request(ctx, request, opts...)
+	if err != nil {
+		return conn, err
+	}
+
+	// Make sure we have a valid expireTime to work with
+	//pathSegment := request.GetConnection().GetNextPathSegment()
+	//expireTime, err := ptypes.Timestamp(pathSegment.GetExpires())
+	//if err != nil {
+	//	return conn, errors.Wrapf(err, "error converting pathSegment.GetExpires() to time.Time: %+v", pathSegment.GetExpires())
+	//}
+
+successVerifyLoop:
+	for {
+		select {
+		case <-u.ctx.Done():
+			u.removeFromExpectedConnections(conn)
+			return conn, errors.Errorf("can't wait for connection %v (segment %v) verification: chain context was canceled", conn.GetId(), conn.GetNextPathSegment())
+		case segment := <-verifySuccessCh:
+			if conn.GetNextPathSegment().Equal(segment) {
+				u.removeFromExpectedConnections(conn)
+				u.addConnectionToMonitor(conn)
+				break successVerifyLoop
+			}
 
-	registerClient := registerClientFunc(ctx)
-	if err == nil && registerClient != nil {
-		registerClient(u.ctx, conn, u.cc)
+			// TODO
+			//  original code used expire time for waiting but it didn't matter because
+			//  originally we created a monitor and received INITIAL_STATE_TRANSFER event which always was instantaneous,
+			//  regardless if there were errors
+			//  now it's unclear how long we should wait
+		//case <-time.After(time.Until(expireTime)):
+		case <-time.After(time.Second):
+			u.removeFromExpectedConnections(conn)
+			return conn, errors.New("timeout expired but we couldn't verify that connection was established")
+		}
 	}
-	return conn, err
+
+	return conn, nil
 }
 
 func (u *healClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	if cachedConn, err := u.removeConnectionFromMonitor(conn); err == nil {
+		conn = cachedConn
+	} else {
+		// TODO log this?
+	}
+
 	return next.Client(ctx).Close(ctx, conn, opts...)
 }
+
+func (u *healClient) listenToConnectionChanges(heal HealRequestFunc, errCh chan error) {
+	monitorClient, err := u.cc.MonitorConnections(u.ctx, &networkservice.MonitorScopeSelector{})
+	if err != nil {
+		errCh <- errors.Wrap(err, "MonitorConnections failed")
+		return
+	}
+	close(errCh)
+
+	// TODO should we really compare paths from the beginning?
+	pathStartsWith := func(c *networkservice.Connection, reference *networkservice.Connection) bool {
```

Comment:
If we use the selector mentioned above, we only need to check `conn.GetPrevPathSegment().Id` instead of full path prefix.